File: code/ex10.java
// DP
// s[i] == 'a' && p[j] == 'a': dp[i][j] = dp[i - 1][j - 1]
// p[j] == '.': dp[i][j] = dp[i - 1][j - 1]
// p[i] == '*': dp[i][j] = dp[i][j - 2] (repete char before zero time)
//                      || dp[i - 1][j] (if p[j - 1] == s[i] || p[j - 1] == '.')

class Solution {
    public boolean isMatch(String s, String p) {
        char[] S = s.toCharArray(), P = p.toCharArray();
        int ns = S.length, np = P.length;
         
        // Init dp array
        boolean[][] dp = new boolean[ns + 1][np + 1];
        dp[0][0] = true;
        
        for (int j = 2; j <= np; j++) {    // If s is empty string
            if (P[j - 1] == '*') {    // If current position is '*'
                dp[0][j] = dp[0][j - 2];
            }
        }
        
        // Build dp array
        for (int i = 1; i <= ns; i++) {
            for (int j = 1; j <= np; ++j) {
                if (S[i - 1] == P[j - 1] || P[j - 1] == '.')
                    dp[i][j] = dp[i - 1][j - 1];
                
                if (P[j - 1] == '*') {
                    dp[i][j] |= dp[i][j - 2];
                    
                    if (j >= 2 && (P[j - 2] == S[i - 1] || P[j - 2] == '.'))
                        dp[i][j] |= dp[i - 1][j];
                }
            }
        }   
        
        return dp[ns][np];
    }
}

File: code/ex10.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int ns = s.length(), np = p.length();
        
        vector<vector<bool>> dp(ns + 1, vector<bool>(np + 1));
        // Init dp array
        dp[0][0] = 1;
        for (int j = 2; j <= np; j++)
            if (p[j - 1] == '*')
                dp[0][j] = dp[0][j - 2];    // Match zero of the element before
        
        // Build dp array
        for (int i = 1; i <= ns; i++) {
            for (int j = 1; j <= np; j++) {
                if (s[i - 1] == p[j - 1] || p[j - 1] == '.')    // Meet with a '.', match previous substring
                    dp[i][j] = dp[i - 1][j - 1];

                else if (p[j - 1] == '*') {    // Meet with a '*'
                    dp[i][j] = dp[i][j] || dp[i][j - 2];    // Match with zero previous char
                    if (j >= 2 && (s[i - 1] == p[j - 2] || p[j - 2] == '.'))    // Match one or more previous chars
                        dp[i][j] = dp[i][j] || dp[i - 1][j];
                }
            }
        }
        
        return dp[ns][np];
    }
};